File: python/dijkstra/priority_queue.py
import math

class PriorityQueue():
  def __init__(self):
    self.contents = []

  def add(self, item, priority):
    self.contents.append( (item, priority) )
    self.heapUp()

  def poll(self):
    contentLen = len(self.contents)
    if contentLen == 0:
      return None
    if contentLen == 1:
      return self.contents.pop()[0]

    item = self.contents[0]
    self.contents[0] = self.contents.pop()
    self.heapDown()
    return item[0]

  # re-ordering
  def heapUp(self):
    current = len(self.contents) - 1

    while self.isParentDisordered(current):
      parentIndex = self.getParentIndex(current)
      self.swap(current, parentIndex)
      current = parentIndex

  def heapDown(self):
    current = 0
    nxt = None

    while self.hasLeft(current):
      if self.isRightLeftOrdered(current):
        nxt = self.getRightIndex(current)
      else:
        nxt = self.getLeftIndex(current)

      if self.isOrdered(self.contents[current], self.contents[nxt]):
        break

      self.swap(current, nxt)
      current = nxt

  # order validation
  def isOrdered(self, above, below):
    return above[1] < below[1]

  def isParentDisordered(self, i):
    return\
      self.hasParent(i) and\
      not self.isOrdered(self.parent(i), self.contents[i])

  def isRightLeftOrdered(self, i):
    return\
       self.hasRight(i) and\
       self.isOrdered(self.right(i), self.left(i))

  # contents of parent, left, right
  def parent(self, i):
    return self.contents[self.getParentIndex(i)]

  def left(self, i):
    return self.contents[self.getLeftIndex(i)]

  def right(self, i):
    return self.contents[self.getRightIndex(i)]

  # presence of parent, left, right
  def hasParent(self, i):
    return self.getParentIndex(i) >= 0

  def hasLeft(self, i):
    return self.getLeftIndex(i) < len(self.contents)

  def hasRight(self, i):
    return self.getRightIndex(i) < len(self.contents)

  # indexes of parent, left right
  def getParentIndex(self, i):
    return int(math.floor((i - 1) / 2))

  def getLeftIndex(self, i):
    return (2 * i) + 1

  def getRightIndex(self, i):
    return (2 * i) + 2

  def swap(self, a, b):
    temp = self.contents[a]
    self.contents[a] = self.contents[b]
    self.contents[b] = temp

  def isEmpty(self):
    return len(self.contents) == 0

  def __str__(self):
    return str(self.contents)

  def __repr__(self):
    return str(self.contents)

File: javascript/dijkstra/priority_queue.js
const PriorityQueueElement = function (key, priority) {
  this.key = key;
  this.priority = priority;
};

const PriorityQueue = function () {
  this.elements = [];
};

PriorityQueue.prototype.add = function (key, priority) {
  this.elements.push(new PriorityQueueElement(key, priority));
  this.heapUp();
};

PriorityQueue.prototype.poll = function () {
  if (this.elements.length == 0) {
    return;
  }
  if (this.elements.length == 1) {
    return this.elements.pop().key;
  }

  const ret = this.elements[0];
  this.elements[0] = this.elements.pop();
  this.heapDown();
  return ret.key;
};

PriorityQueue.prototype.heapUp = function () {
  let current = this.elements.length - 1;
  while (this.isParentDisordered(current)) {
    const parentIndex = this.getParentIndex(current);
    this.swap(current, parentIndex);
    current = parentIndex;
  }
};

PriorityQueue.prototype.heapDown = function () {
  let current = 0;
  let nxt = 0;

  while (this.hasLeft(current)) {
    if (this.isRightLeftOrdered(current)) {
      nxt = this.getRightIndex(current);
    } else {
      nxt = this.getLeftIndex(current);
    }

    if (this.isOrdered(current, nxt)) {
      return;
    }

    this.swap(current, nxt);
    current = nxt;
  }
};

PriorityQueue.prototype.isOrdered = function (above, below) {
  return this.elements[above].priority < this.elements[below].priority;
};
PriorityQueue.prototype.isParentDisordered = function (i) {
  return this.hasParent(i) && !this.isOrdered(this.getParentIndex(i), i);
};
PriorityQueue.prototype.isRightLeftOrdered = function (i) {
  return this.hasRight(i) && this.isOrdered(this.getRightIndex(i), this.getLeftIndex(i));
};
PriorityQueue.prototype.parent = function (i) {
  return this.elements[this.getParentIndex(i)];
};
PriorityQueue.prototype.left = function (i) {
  return this.elements[this.getLeftIndex(i)];
};
PriorityQueue.prototype.right = function (i) {
  return this.elements[this.getRightIndex(i)];
};
PriorityQueue.prototype.hasParent = function (i) {
  return this.getParentIndex(i) >= 0;
};
PriorityQueue.prototype.hasLeft = function (i) {
  return this.getLeftIndex(i) < this.elements.length;
};
PriorityQueue.prototype.hasRight = function (i) {
  return this.getRightIndex(i) < this.elements.length;
};
PriorityQueue.prototype.getParentIndex = function (i) {
  return Math.floor((i - 1) / 2);
};
PriorityQueue.prototype.getLeftIndex = function (i) {
  return (i * 2) + 1;
};
PriorityQueue.prototype.getRightIndex = function (i) {
  return (i * 2) + 2;
};
PriorityQueue.prototype.swap = function (aIndex, bIndex) {
  const temp = this.elements[aIndex];
  this.elements[aIndex] = this.elements[bIndex];
  this.elements[bIndex] = temp;
};
PriorityQueue.prototype.isEmpty = function () {
  return this.elements.length == 0;
};

module.exports = PriorityQueue;